### ISSUE TITLE
More Documentation Around the CLI --require

### ISSUE THREAD

**Issue by ericadohring** [mbj/mutant#1056]
Posted a q on stack overflow. More documentation would be super helpful to make this more clear
https://stackoverflow.com/questions/64094367/how-to-pass-in-arguments-to-mutant-rspec

**Comment by dogweather** [comment]
Yep, the library is unusable as is: a 40-second delay just to view the meager `--help` docs. And then there are no examples of invocations, and most options aren't documented. Lots of playing around is required to get a successful run on, e.g., a Rails app. And at 40 second wait per attempt? No go.

**Comment by pawelpacana** [comment]
> the library is unusable

Perhaps we have a different background. The library has been of tremendous value for me over the years. There are at least a few public, open-source projects using it which can serve as an additional example to [docs](https://github.com/mbj/mutant/tree/master/docs).

> a 40-second delay just to view the meager --help docs

That would be quite suspicious, indeed:

```
$ time bundle exec mutant --help                                                                                                    495ms 
usage: mutant [options] MATCH_EXPRESSION ...
Environment:
        --zombie                     Run mutant zombified
    -I, --include DIRECTORY          Add DIRECTORY to $LOAD_PATH
    -r, --require NAME               Require file with NAME
    -j, --jobs NUMBER                Number of kill jobs. Defaults to number of processors.

Options:
        --use INTEGRATION            Use INTEGRATION to kill mutations
        --ignore-subject EXPRESSION  Ignore subjects that match EXPRESSION as prefix
        --since REVISION             Only select subjects touched since REVISION
        --fail-fast                  Fail fast
        --version                    Print mutants version
    -h, --help                       Show this message

________________________________________________________
Executed in  492,57 millis    fish           external 
   usr time  346,00 millis   72,00 micros  345,93 millis 
   sys time  133,21 millis  361,00 micros  132,84 millis 
```

**Comment by mbj** [comment]
@pawelpacana It could be that @dogweather is seeing a case where he is trying the `--help` on an unlicensed repository.

And this is a valid concern. I'll reproduce this case and make sure that the 40s "nudge" timer is only active for actually running mutations.

**Comment by mbj** [comment]
Yeah, so `--help` on an unlicensed repository indeed causes the 40s nudging. I'm actually fine with that behavior. Its unlicensed usage anyway.

@dogweather I think you know whats going on. You are using mutant on an unlicensed repository and get a 40s pause for that.

I'm happy to provide you with a free OSS license, just mail me a public repository URL you wish to use mutant on.

Other than that: I think mutant is not as well documented as it could be. Still we have a great community here, should you have non abstract questions on how to run it feel free to voice them here.

**Comment by dogweather** [comment]
@mbj Thanks for the reply. I'm sorry about my sharply worded comment — it was frustrating to realize that every interaction would incur the 40s penalty.

But I have **no** problem with wanting to earn money from one's work, and your innovative approach is interesting to see. I also think your pricing is reasonable, if it delivers what I _imagine_ it does. IMO your solution "only" needs the adjustment so that, without paying money, a potential customer can see the value they'd be getting.

**Comment by dogweather** [comment]
Non-abstract questions:

* What are examples of running mutant on a Rails codebase?
* What does the solitary `--` in one of the doc'd invocations do?
* What, precisely, does the final CLI argument match?

**Comment by mbj** [comment]
> it was frustrating to realize that every interaction incurs the 40s penalty, including viewing the help docs.

no worries. I hope I measured my response right. The fact you are still interacting seems to confirm this. Peace ;)
Writing mutant is very frustrating in itself as Ruby is such a hostile environment for tooling. So we are sorta experiencing the same through different filters.

Okay on the topic.

> IMO your solution "only" needs the adjustment so that, without paying money, a potential customer can see the product's value.

Mutant is free to use on any public repository that has an opensource license. This free usage *still* requires a license key whic is easy to get: simply mail me.

If you wanted to spare setup there are some public examples that have mutant pre-installed. And have a license key already. As opensource license keys license the repository NOT the user they are ready to go.

* https://github.com/mbj/mutant - yes mutant is self hosting, NOT the best place to start.
* https://github.com/mbj/auom - very easy domain (compared) as there is no IO, part of the mutant self test corpus. And has rspec and minitest coverage. (Mintest being 10x faster than rspec under MT).
* https://github.com/mbj/unparser - self hosting example, NOT a good place to start
* lots of other repos under `mbj/*` but mostly they are all part of mutants dependencies.

Now more interesting ones:

* https://github.com/schirp-dso/mutant-on-rails-workshop - a workshop I've ran multiple times on a reasonably complex domain. Caveats are that I only use sqllite there which does not explore the rails under concurrency DB side of things.
* https://github.com/arkency/aggregates
* https://github.com/RailsEventStore/rails_event_store <- a very good example

The history of this project is:

* Started out as "underground" OSS. Where I actively did NOT look for wider audience. I never intentionally documented it and keept the knowledge in the head / history of the datamapper 1.0 / 2.0 IRC channel. Simply because maintaining tool with wide usage under OSS economy is NOT my life goal.
* Used mutant as "secret weapon" for Ruby/Rails consulting and greatly leveled up its accurancy. Only used actively by ~10 people I all know personally.
* Quit full time ruby consulting, and have this engine that is too useful to let die, and hence transitioning to a commerical offering. Now have to backfill ~10y of making a nice on ramp. While spending the limited time I have on mutant on also serving commercial clients that knew me from consulting and their advanced feature requests.

I'm still very biased towards the needs of *existing* power users that are active subscribers. The backfilling of information for new users is lacking ~7y behind the evolution of the tool.*.

Basically mutant is: Very useful for people that know on how to use it. Super hostile to everyone else. I'm well aware, but the former group actively pays me. And the latter group only potentially pays me. Hence I'm biased for the former group. I wish to fix that but thats hard.

Let me flush this edit buffer and post it. Than react to the questions you added meanwhile.

**Comment by mbj** [comment]
> I looked at your community support options, and Slack is too much friction and too ephemeral for me as a _potential_ customer. A more permanent record like a Discourse instance would be more helpful because I could see Q&A's.

We (the wider mutation testing community) is actively discussing to move our shared mutation testing slack channel to discourse. I've buy in from some other tool authors.

**Comment by mbj** [comment]
Let me elaborate the example you posted:

```
bundle exec mutant
```
just starts mutant.

```
--include lib
```

the same thing rubyies --include does, but controlled by mutant to be executed at the "right time".

```
--require auom
```

the same thing rubies --require does, but controlled by mutant to be executed at the "right time".

```
--use rspec
```

instructs mutant to use the rspec integration / rspec related auto-discovery

```
-- 
```

argument separator https://unix.stackexchange.com/questions/147143/when-and-how-was-the-double-dash-introduced-as-an-end-of-options-delimiter NOT specific to mutant at all. Just good practise to make sure that the optparse library stops interpreting things as arguments. I have it in all my scripts (And examples) that potentially construct CLIs dynamically to educate people to do the "right thing" (tm)

```
'AUOM*'
```

Subject match expression. Quoted to avoid shell expansion. Subject match expressions telll mutant "what" to mutate. This expression basically means: Mutate everything inside the `AUOM` namespace, recursively. There are other match expressions. You can target individual classes/modules (non recursive): `AUOM`, a specific instance method `AUOM#some_method` and singleton methods `AUOM.some_method`. These affect what is included in mutant runs. And depending on your execution scenario (CI, manual hunting, code review) you wish to use the right kind of expression for a fast or verbose feedback.

There are other options that affect subject selection. I'll not go into them here. Still these other match expression examples explain why I generally quote things in my examples. As `AUOM#some_method` could easily be interpreted as a shell comment.

You can get the same semantics via:

```
bundle exec mutant --include lib --require auom --use rspec AUOM*
```

but its less safe for scripting. Sorry the `--` throws you off.

And to reduce redundancy you can create a mutant config file (undocumented ; ) to move the `--include` and `--require` out of recurring CLI.

So if you are interested in mutation testing, given the sad state of documentation of mutant iself: I'm happy to shedule a call. Takes ~30min and we can go over any OSS project of your choice, from scratch of from the ones I already have set-up.

This is my primary sales and feedback vector right now. I'm aware it does not scale well but I've yet to find the time to do the upfront investment to make "standalone resources". In fact I'm actively searching to contract someone to make it for me.

**Comment by mbj** [comment]
@dogweather BTW we have a discord channel now that replaces the Slack channel, its more accessible.

**Comment by mbj** [comment]
I'm closing this as I think this issue does not represent a distinct actionable.
